crates/bevy_ecs/src/observer/mod.rs

```diff
@@ -0,0 +1,599 @@
+            let world = world.as_unsafe_world_cell();
+            // SAFETY: There are no outstanding world references
+            world.increment_event_id();
+            let observers = world.observers();
+            let Some(observers) = observers.try_get_observers(event) else {
+                return;
+            };
+            // SAFETY: The only outstanding reference to world is `observers`
+            (world.into_deferred(), observers)
+        };
+
+        let mut trigger_observer = |(&observer, runner): (&Entity, &ObserverRunner)| {
+            (runner)(
+                world.reborrow(),
+                ObserverTrigger {
+                    observer,
+                    event,
+                    source,
+                },
+                data.into(),
+            );
+        };
+
+        // Trigger observers listening for any kind of this event
+        observers.map.iter().for_each(&mut trigger_observer);
+
+        // Trigger entity observers listening for this kind of event
+        if source != Entity::PLACEHOLDER {
+            if let Some(map) = observers.entity_observers.get(&source) {
+                map.iter().for_each(&mut trigger_observer);
+            }
+        }
+
+        // Trigger observers listening to this event targeting a specific component
+        components.for_each(|id| {
+            if let Some(component_observers) = observers.component_observers.get(&id) {
+                component_observers
+                    .map
+                    .iter()
+                    .for_each(&mut trigger_observer);
+
+                if source != Entity::PLACEHOLDER {
+                    if let Some(map) = component_observers.entity_map.get(&source) {
+                        map.iter().for_each(&mut trigger_observer);
+                    }
+                }
+            }
+        });
+    }
+
+    pub(crate) fn is_archetype_cached(event: ComponentId) -> Option<ArchetypeFlags> {
+        match event {
+            ON_ADD => Some(ArchetypeFlags::ON_ADD_OBSERVER),
+            ON_INSERT => Some(ArchetypeFlags::ON_INSERT_OBSERVER),
+            ON_REMOVE => Some(ArchetypeFlags::ON_REMOVE_OBSERVER),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn update_archetype_flags(
+        &self,
+        component_id: ComponentId,
+        flags: &mut ArchetypeFlags,
+    ) {
+        if self.on_add.component_observers.contains_key(&component_id) {
+            flags.insert(ArchetypeFlags::ON_ADD_OBSERVER);
+        }
+        if self
+            .on_insert
+            .component_observers
+            .contains_key(&component_id)
+        {
+            flags.insert(ArchetypeFlags::ON_INSERT_OBSERVER);
+        }
+        if self
+            .on_remove
+            .component_observers
+            .contains_key(&component_id)
+        {
+            flags.insert(ArchetypeFlags::ON_REMOVE_OBSERVER);
+        }
+    }
+}
+
+impl World {
+    /// Construct an [`ObserverBuilder`]
```

Comment:
This needs a line explaining when to use this vs the `observer` method (presumably it's when you need elaborate config).